### PR TITLE
Python more numpy

### PR DESCRIPTION
**Pull request by wrongbad**
Note: this branch is on top of WarpBatch commit - that will disappear when that other PR merges.

Changed `hull_points` to accept numpy point arrays. Also removed several redundant numpy shape checks - these shape constraints were already enforced by the shape template parameters.

`sdf` / `level_set` could benefit from numpy batching but requires deeper refactoring (#598)

`set_properties` also could benefit, and I'd like to look into it more here.

That's all that I see. Let me know if I missed anything.

While I was auditing the API, I noticed something else needing cleanup. Many functions need  a single vector/point, but the convention is all over the map. Some take 3 floats, some take a tuple, some take a `glm::vec3`, some take a numpy(3), and several have overloads to support some random subset of those. We should probably converge on 1 way. My preference would be tuple, but I'm curious to hear other opinions.

### CODE REVIEW

**Comment by pca006132** [comment]
There is a caster for `glm::vec3` that can accept anything that is a sequence with 3 elements, e.g. list, tuple and numpy arrays.

**Comment by pca006132** [comment]
I think there are also some other tiny fixes that can be added into this PR.

Missing `\n` in line 371:
https://github.com/elalish/manifold/blob/eba866d266f1954f81211c71d9f5329477606e5e/bindings/python/manifold3d.cpp#L368-L375

This should be an int instead of a double:
https://github.com/elalish/manifold/blob/eba866d266f1954f81211c71d9f5329477606e5e/bindings/python/manifold3d.cpp#L997

**Comment by wrongbad** [comment]
> Not sure if arguments accepting numpy arrays can accept normal python list.

A quick test suggests the answer is no. I guess the auto casting approach has the best upside. I'll tinker with it some more.

**Comment by wrongbad** [comment]
I removed almost all of the lambda functions in the bindings. A few remain for overload-resolution, dealing with stray types like `Rect` or `Smoothness`, or doing something non-trivial.

**Comment by wrongbad** [comment]
I also removed some overloads like `cube(float, float, float)` when `cube(vec3)` exists. I think right now a primary obstacle for maintainability and cleanliness is the sheer quantity of code in the manifold3d.cpp, so anything that reduces that code footprint is good.

**Comment by pca006132** [comment]
Thanks for the cleanup. I think much of the code comes from docstrings, it would be nice if we can find a way to unify the documentation for the main C++ library and the 3 different bindings, but I guess this is a hard problem (e.g. name inconsistencies, specifics for certain binding).

**Comment by elalish** [comment]
This looks awesome, thanks! I always appreciate cleanups that will ease maintenance going forward. I'll review more closely after the other PR is merged and this comes out of draft.

**Comment by pca006132** [comment]
I think you probably need to push a dummy commit to trigger the CI build. We should fix this CI problem later...

File: bindings/python/manifold3d.cpp
Comment (pca006132) [inline bindings/python/manifold3d.cpp:256]: issues with space characters here.

File: bindings/python/manifold3d.cpp
Comment (pca006132) [inline bindings/python/manifold3d.cpp:303]: parameter name is different.

File: bindings/python/manifold3d.cpp
Comment (pca006132) [inline bindings/python/manifold3d.cpp:349]: thinking about it, we may want documentation about the type signature of the function. same for `warp` above and `set_properties` below.

File: bindings/python/manifold3d.cpp
Comment (pca006132) [inline bindings/python/manifold3d.cpp:502]: parameter name

File: bindings/python/manifold3d.cpp
Comment (pca006132) [inline bindings/python/manifold3d.cpp:513]: parameter name

File: bindings/python/manifold3d.cpp
Comment (pca006132) [inline bindings/python/manifold3d.cpp:533]: we should probably state that this is a breaking change, i.e. property vs method.

File: bindings/python/manifold3d.cpp
Comment (pca006132) [inline bindings/python/manifold3d.cpp:545]: I'm thinking if we want to make this numpy-friendly, maybe by splitting it into an integer index array and smoothness array.

File: bindings/python/manifold3d.cpp
Comment (pca006132) [inline bindings/python/manifold3d.cpp:605]: need to change `radius_low` and `radius_high` as well.

File: bindings/python/manifold3d.cpp
Comment (pca006132) [inline bindings/python/manifold3d.cpp:866]: `m` vs `mat`

File: bindings/python/manifold3d.cpp
Comment (pca006132) [inline bindings/python/manifold3d.cpp:975]: `dims` vs `size`

**Review by pca006132 — CHANGES_REQUESTED** [review]
some nitpicks.

btw, not sure if we want `Manifold const &self`, I think we typically write `const Manifold &self`, and order does not matter for references (`T&` is like `T const *` anyway).

**Comment by wrongbad** [comment]
What do you think about `bounding_box` -> `bounding_box()` and `get_surface_area()` -> `surface_area()` and `get_volume()` -> `volume()`?

**Comment by pca006132** [comment]
I think every methods on `Manifold` may require evaluating the underlying CSG graph, so they can be really expensive and should probably be methods instead of attributes. Removing the `get_` prefix seems fine to me.

File: bindings/python/manifold3d.cpp
Comment (wrongbad) [reply bindings/python/manifold3d.cpp]: Makes sense to me. Done

File: bindings/python/examples/all_apis.py
Comment (elalish) [inline bindings/python/examples/all_apis.py:47]: Thanks - this big basic test file is a great idea. Nice as an API index too.

**Review by elalish — APPROVED** [review]
That's a lot of great clean-up. LGTM; I'll leave it to @pca006132 for final approval since I'm no Python master.

**Comment by elalish** [comment]
Looks like a pretty simple error the python test caught, but are we only running Python tests on some of our builds? Is that intentional @pca006132?

**Comment by pca006132** [comment]
Forgot if this is intentional or not, we can just enable it for every build. We can also enable all tests even for draft PR, so we will not get skipped tests when we change the state from draft to ready for review.

I think I will wait for @wrongbad to signal this ready for review, not sure if he is done with it now.

**Comment by wrongbad** [comment]
At this point I'm just hoping to get all the tests passing, and resolve any review concerns. I'll look at generating doc-strings in a follow-up.

**Comment by pca006132** [comment]
I will have a look at it tmr. We can leave the documentation generation thing later. I feel that it should not be too complicated, but may require parsing the documentation to get a list of parameters for substitution. Can probably be done together with the python stub.

File: bindings/python/manifold3d.cpp
Comment (pca006132) [inline bindings/python/manifold3d.cpp:141]: This is surprising. Just curious, any reference on this?

**Review by pca006132 — APPROVED** [review]
I think I still see some inconsistencies in parameter name for the docstring, but I guess we can pass this for now and embrace automation :)

File: bindings/python/manifold3d.cpp
Comment (wrongbad) [reply bindings/python/manifold3d.cpp:141]: So this function is `noexcept` and if you call `nb::len` on a `float` it throws, and terminates the program. So I looked into source for `nb::len` and found this function which gives a softer fallback mode.

**Comment by pca006132** [comment]
So I guess this can be merged?

**Pull request merged**
